trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/MainTranslator.java

```diff
@@ -237,6 +239,18 @@ private CHandlerTranslationResult performMainRun(final TranslationSettings trans
 		final MainDispatcher mainDispatcher = new MainDispatcher(mLogger, witness, locationFactory, typeHandler,
 				mainCHandler, ppHandler, acslHandler);
 
+		if (witness != null) {
+			// Add the declaration of global ghost variables
+			for (final var ghost : witness.getGlobalDeclarations()) {
+				final ExpressionResult res = ghost.getInitializationResult(mainDispatcher);
+				staticObjectsHandler.addStatementsForUltimateInit(res.getStatements());
+				for (final var d : res.getDeclarations()) {
+					staticObjectsHandler.addGlobalVarDeclarationWithoutCDeclaration((VariableDeclaration) d);
```

Comment:
question: what are these declarations? how are they related to the declaration added after the loop? should all of them be global?

Reply:
Previously `res` itself contained the declarations. This was however changed, this loop can be removed.

Reply:
Can the initialization require declarations of other variables? e.g. auxvars? (though they should probably not be global)

Reply:
Currently this is not possible, since we (luckily) restricted the ghost updates, s.t. they can only have a primitive type and their initial value can only be a constant expression. If we wanted to be less restrictive, we probably need to adapt this anway (as it seems, we cannot add aux-vars or local declarations to the init-procedure here).

Reply:
I think we said in the schema that the expression may be any expression over global variables, right @bahnwaerter?

Reply:
(side-effect-free and without undefined behaviour, of course)

Reply:
> The only variables allowed in this expression are the global variables of the program.
> The expression is evaluated after global variables declared in the program are initialized.

Okay, I see, then there might be aux-vars involved, e.g. if a ghost variable is initialized by a read of a global array.
But in that case the second sentence is problematic. Currently we initialize the ghost variables (in ULTIMATE.init) before the program. To match the specification, we would need to initialize the ghost variables at the very end, i.e., between dispatching the actual program, but before calling `PostProcessor` (where the ULTIMATE.init procedure is actually build). 
I tried to implement this in f6123d9, but I am not sure, if this is a nice solution...